draco/asp/helpers.lp

```diff
@@ -106,24 +116,96 @@ attribute(is_c_c,M) :-
 attribute(is_c_d,M) :-
     entity(mark,_,M),
-    attribute((mark,channel),M,x),
-    attribute((mark,channel),M,y),
-    not attribute(mark_channel_discrete_or_binned,M,x),
-    attribute(mark_channel_discrete_or_binned,M,y).
+    not attribute(mark_channel_cont,M,x),
+    attribute(mark_channel_cont,M,y).
 
 % @helper(is_d_d) Discrete by discrete.
 attribute(is_d_d,M) :-
-    entity(mark,_,M),
     attribute((mark,channel),M,x),
     attribute((mark,channel),M,y),
     attribute(mark_channel_discrete_or_binned,M,x),
     attribute(mark_channel_discrete_or_binned,M,y).
+
+% @helper(non_pos_unaggregated) Non-positional channels are unaggregated.
+attribute(non_pos_unaggregated,M) :-
+    entity(encoding,M,E),
+    attribute((encoding,channel),E,C),
+    domain(non_positional,C),
+    not attribute((encoding,aggregate),E,_).
+
+% @helper(no_overlap) The continuous variable is a measure (it is aggregated) and all other non-positional channels are aggregated, or we use stack.
+attribute(no_overlap,M) :-
+    attribute(is_c_d,M),
+    attribute(mark_encoding_cont,M,E),
+    attribute((encoding,channel),E,(x;y)),
+    attribute((encoding,aggregate),E,_),
+    not attribute(non_pos_unaggregated,M).
+attribute(no_overlap,M) :-
+    attribute(is_c_d,M),
+    entity(encoding,M,E),
+    attribute((encoding,stack),E,_).
+attribute(no_overlap,M) :-
+    attribute(is_c_d,M),
+    attribute(number_rows,root,N),
+    attribute(discrete_size,M,N).
+attribute(no_overlap,M) :-
+    attribute(is_d_d,M),
+    not attribute(non_pos_unaggregated,M).
+attribute(no_overlap,M) :-
+    attribute(is_d_d,M),
+    attribute(number_rows,root,N1),
+    attribute(discrete_size,M,N2),
+    N1 <= N2.
+
+% @helper(overlap) We definitely overlap if the data size > discrete size.
+attribute(overlap,M) :-
+    attribute(is_c_d,M),
+    not attribute(no_overlap,M),
+    attribute(number_rows,root,N1),
+    attribute(discrete_size,M,N2),
+    N1 > N2.
+attribute(overlap,M) :-
+    entity(mark,V,M),
+    attribute(is_d_d,M),
+    not attribute(no_overlap,M),
+    not entity(facet,V,_),
+    attribute(number_rows,root,N1),
+    attribute(discrete_size,M,N2),
+    N1 > N2.
+
+% @helper(discrete_size) The size of the discrete positional encoding.
+attribute(discrete_size,M,N) :-
+    attribute(is_c_d,M),
+    attribute(x_y_cardinality,M,_,N).
+attribute(discrete_size,M,1) :-
+    attribute(is_c_d,M),
+    attribute(mark_encoding_cont,M,E),
+    attribute((encoding,channel),E,x),
+    not attribute((mark,channel),M,y).
+attribute(discrete_size,M,1) :-
+    attribute(is_c_d,M),
+    attribute(mark_encoding_cont,M,E),
+    attribute((encoding,channel),E,y),
+    not attribute((mark,channel),M,x).
+attribute(discrete_size,M,N) :-
+    attribute(is_d_d,M),
+    attribute(x_y_cardinality,M,x,NX),
+    attribute(x_y_cardinality,M,y,NY),
+    N = NX*NY.
+
+% @helper(x_y_cardinality) to go from quadratic to linear number of grounding
+attribute(x_y_cardinality,M,x,N) :-
```

Comment:
Since x and y is not consistent here, should we have `attribute(x_cardinality,M,N)` or `attribute(cardinality,M,x,N)`?